File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Benchmarks/Services/TTSBenchmarkProvider.swift
//
//  TTSBenchmarkProvider.swift
//  RunAnywhereAI
//
//  Benchmarks TTS synthesis with short and medium text inputs.
//

import Foundation
import RunAnywhere

struct TTSBenchmarkProvider: BenchmarkScenarioProvider {

    let category: BenchmarkCategory = .tts

    func scenarios() -> [BenchmarkScenario] {
        [
            BenchmarkScenario(name: "Short Text", category: .tts, parameters: ["length": "short"]),
            BenchmarkScenario(name: "Medium Text", category: .tts, parameters: ["length": "medium"]),
        ]
    }

    func execute(
        scenario: BenchmarkScenario,
        model: ModelInfo
    ) async throws -> BenchmarkMetrics {
        var metrics = BenchmarkMetrics()

        let text: String
        switch scenario.parameters?["length"] {
        case "short":
            text = "Hello, this is a test."
        default:
            text = "The quick brown fox jumps over the lazy dog. Machine learning models can generate speech from text with remarkable quality and natural intonation."
        }

        let memBefore = SyntheticInputGenerator.availableMemoryBytes()

        // Load
        let loadStart = Date()
        try await RunAnywhere.loadTTSModel(model.id)
        metrics.loadTimeMs = Date().timeIntervalSince(loadStart) * 1000

        do {
            // Synthesize (not speak)
            let benchStart = Date()
            let options = TTSOptions()
            let result = try await RunAnywhere.synthesize(text, options: options)
            metrics.endToEndLatencyMs = Date().timeIntervalSince(benchStart) * 1000

            // processingTime is in seconds, convert to ms-context
            metrics.audioDurationSeconds = result.duration
            metrics.charactersProcessed = result.metadata.characterCount

            let memAfter = SyntheticInputGenerator.availableMemoryBytes()
            metrics.memoryDeltaBytes = memBefore - memAfter

            try? await RunAnywhere.unloadTTSVoice()
            return metrics
        } catch {
            try? await RunAnywhere.unloadTTSVoice()
            throw error
        }
    }
}

File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/benchmarks/services/TTSBenchmarkProvider.kt
package com.runanywhere.runanywhereai.presentation.benchmarks.services

import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkCategory
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkDeviceInfo
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkMetrics
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkScenario
import com.runanywhere.runanywhereai.presentation.benchmarks.utilities.SyntheticInputGenerator
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.Models.ModelInfo
import com.runanywhere.sdk.public.extensions.TTS.TTSOptions
import com.runanywhere.sdk.public.extensions.loadTTSVoice
import com.runanywhere.sdk.public.extensions.synthesize
import com.runanywhere.sdk.public.extensions.unloadTTSVoice
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.withContext

/**
 * Benchmarks TTS synthesis with short and medium text inputs.
 * Matches iOS TTSBenchmarkProvider exactly.
 */
class TTSBenchmarkProvider : BenchmarkScenarioProvider {

    override val category: BenchmarkCategory = BenchmarkCategory.TTS

    override fun scenarios(): List<BenchmarkScenario> = listOf(
        BenchmarkScenario(name = "Short Text", category = BenchmarkCategory.TTS),
        BenchmarkScenario(name = "Medium Text", category = BenchmarkCategory.TTS),
    )

    override suspend fun execute(
        scenario: BenchmarkScenario,
        model: ModelInfo,
        deviceInfo: BenchmarkDeviceInfo,
    ): BenchmarkMetrics {
        val text = if (scenario.name.contains("Short")) {
            "Hello, this is a test."
        } else {
            "The quick brown fox jumps over the lazy dog. Machine learning models can generate speech from text with remarkable quality and natural intonation."
        }

        val memBefore = SyntheticInputGenerator.availableMemoryBytes()

        // Load (TTS uses "voice" API on Kotlin)
        val loadStart = System.nanoTime()
        RunAnywhere.loadTTSVoice(model.id)
        val loadTimeMs = (System.nanoTime() - loadStart) / 1_000_000.0

        try {
            // Synthesize (not speak)
            val benchStart = System.nanoTime()
            val options = TTSOptions()
            val result = RunAnywhere.synthesize(text, options)
            val endToEndMs = (System.nanoTime() - benchStart) / 1_000_000.0

            val memAfter = SyntheticInputGenerator.availableMemoryBytes()

            return BenchmarkMetrics(
                endToEndLatencyMs = endToEndMs,
                loadTimeMs = loadTimeMs,
                memoryDeltaBytes = memBefore - memAfter,
                audioDurationSeconds = result.duration,
                charactersProcessed = result.metadata.characterCount,
            )
        } finally {
            withContext(NonCancellable) {
                try {
                    RunAnywhere.unloadTTSVoice()
                } catch (_: Exception) {
                }
            }
        }
    }
}
